File: 08-安全/03-应用安全与供应链/供应链完整性/provenance_verify.py
#!/usr/bin/env python3
"""SLSA v1.0 来源证明（Provenance）的生成与验证。

Provenance 回答的是「**这个制品是在哪里、何时、如何被生产出来的**」，
它是 in-toto attestation 框架里的一种 predicate：

    {
      "_type": "https://in-toto.io/Statement/v1",
      "subject": [ ResourceDescriptor, ... ],      # 构建输出（顶层字段，不在 predicate 里）
      "predicateType": "https://slsa.dev/provenance/v1",
      "predicate": {
        "buildDefinition": { buildType, externalParameters, internalParameters, resolvedDependencies },
        "runDetails": { builder{id,...}, metadata{invocationId, startedOn, finishedOn}, byproducts }
      }
    }

本 demo 的重点**不是生成**，而是**验证**：规范里那些 MUST / SHOULD 到底在防什么。
每条验证规则对应一个可运行的攻击场景。
"""

import base64
import hashlib
import hmac
import json
from typing import Dict, List, Optional, Tuple

PREDICATE_TYPE = "https://slsa.dev/provenance/v1"
STATEMENT_TYPE = "https://in-toto.io/Statement/v1"

# builder.id -> 该构建平台声明的 SLSA Build level
# 规范原话：builder.id "is intended to be the sole determiner of the SLSA Build level"
BUILDER_LEVELS: Dict[str, int] = {
    "https://github.com/actions/runner/github-hosted": 2,
    "https://cloudbuild.googleapis.com/GoogleCloudBuild": 3,
}

# 消费者**只接受特定"签名者—构建者"配对**（Consumers MUST accept only specific
# signer-builder pairs）：GitHub 可以为 GitHub Actions 签名，但不能为 Google Cloud Build 签名。
ACCEPTED_PAIRS = {
    ("github", "https://github.com/actions/runner/github-hosted"),
    ("google", "https://cloudbuild.googleapis.com/GoogleCloudBuild"),
}

# 我们对「正常的 externalParameters」的预期。规范：这些值**不可信**，必须下游验证，
# 且验证者 SHOULD 拒绝其中未识别/意外的字段。
EXPECTED_EXTERNAL_KEYS = {"repository", "ref"}


# ------------------------------------------------------------------ 工具

def canonical(obj) -> bytes:
    """确定性序列化（真实 DSSE 用类似的规范化，签名才不会被键序影响）。"""
    return json.dumps(obj, sort_keys=True, separators=(",", ":")).encode()


def sha256_hex(b: bytes) -> str:
    return hashlib.sha256(b).hexdigest()


def sign(statement: dict, key: bytes, key_id: str) -> dict:
    """用 HMAC-SHA256 模拟签名（不引入第三方依赖；真实场景是 DSSE + 公钥签名）。"""
    payload = base64.b64encode(canonical(statement)).decode()
    sig = hmac.new(key, canonical(statement), hashlib.sha256).hexdigest()
    return {"payload": payload, "keyId": key_id, "signature": sig}


def open_envelope(env: dict, key: bytes) -> Optional[dict]:
    if not hmac.compare_digest(
            hmac.new(key, base64.b64decode(env["payload"]), hashlib.sha256).hexdigest(),
            env["signature"]):
        return None
    return json.loads(base64.b64decode(env["payload"]))


def make_provenance(builder_id: str, subject_name: str, artifact: bytes,
                    external: dict, internal: Optional[dict] = None,
                    deps: Optional[List[dict]] = None,
                    extra_predicate: Optional[dict] = None,
                    invocation: str = "inv-0001") -> dict:
    p = {
        "_type": STATEMENT_TYPE,
        "subject": [{"name": subject_name, "digest": {"sha256": sha256_hex(artifact)}}],
        "predicateType": PREDICATE_TYPE,
        "predicate": {
            "buildDefinition": {
                "buildType": "https://slsa-framework.github.io/github-actions-buildtypes/workflow/v1",
                "externalParameters": dict(external),
                "internalParameters": internal or {},
                "resolvedDependencies": deps or [],
            },
            "runDetails": {
                "builder": {"id": builder_id},
                "metadata": {
                    "invocationId": invocation,
                    "startedOn": "2026-09-18T10:00:00Z",
                    "finishedOn": "2026-09-18T10:04:00Z",
                },
                "byproducts": [],
            },
        },
    }
    if extra_predicate:
        p["predicate"].update(extra_predicate)
    return p


# ------------------------------------------------------------------ 验证

def verify(env: dict, key: bytes, artifact: bytes, min_level: int = 2) -> Tuple[str, List[str]]:
    """返回 (decision, reasons)。decision ∈ {"ALLOW", "DENY"}。"""
    reasons: List[str] = []

    stmt = open_envelope(env, key)
    if stmt is None:
        return "DENY", ["签名校验失败"]

    # 1) subject 摘要必须等于手上制品的摘要 —— 防「构建后又换了包」
    got = stmt["subject"][0]["digest"]["sha256"]
    want = sha256_hex(artifact)
    if not hmac.compare_digest(got, want):
        return "DENY", ["subject 摘要与制品不匹配（制品在构建后被替换）"]

    if stmt["predicateType"] != PREDICATE_TYPE:
        return "DENY", ["predicateType 不是 SLSA provenance v1"]

    pred = stmt["predicate"]
    bd = pred.get("buildDefinition", {})
    rd = pred.get("runDetails", {})
    builder_id = rd.get("builder", {}).get("id", "")

    # 2) 签名者—构建者配对。builder 与 signer 是分开的角色，不能混。
    if (env["keyId"], builder_id) not in ACCEPTED_PAIRS:
        return "DENY", ["不接受该 signer-builder 配对: signer=%s builder=%s"
                        % (env["keyId"], builder_id)]

    # 3) builder.id 是 SLSA Build level 的**唯一**决定因素
    level = BUILDER_LEVELS.get(builder_id, 0)
    if level < min_level:
        msg = "builder.id 声明的级别 %d 低于要求 %d" % (level, min_level)
        # 扩展字段不得改变其它字段的含义 → 自称的级别不作数
        if "x_slsaBuildLevel" in pred:
            msg += "；扩展字段 x_slsaBuildLevel=%r 不得用于提升级别（单调性原则）" \
                   % pred["x_slsaBuildLevel"]
        return "DENY", [msg]

    # 4) externalParameters 不可信 → 必须与预期一致，且拒绝未识别字段
    ext = bd.get("externalParameters", {})
    unexpected = set(ext) - EXPECTED_EXTERNAL_KEYS
    if unexpected:
        return "DENY", ["externalParameters 出现未预期字段: %s" % sorted(unexpected)]

    # 5) internalParameters 由可信平台设置 → **不需要**验证（规范明说）
    reasons.append("internalParameters 由受信任平台设置，按规范不校验")

    # 6) 未识别的顶层字段必须忽略（Consumers MUST ignore unrecognized fields）
    known_top = {"_type", "subject", "predicateType", "predicate"}
    ignored = sorted(set(stmt) - known_top)
    if ignored:
        reasons.append("忽略未识别顶层字段: %s" % ignored)
    known_pred = {"buildDefinition", "runDetails"}
    ignored_p = sorted(set(pred) - known_pred)
    if ignored_p:
        reasons.append("忽略未识别 predicate 字段（扩展）: %s" % ignored_p)
        # 扩展的单调性原则：忽略扩展不应把 DENY 变成 ALLOW。
        # 因此**任何**策略都不得依赖扩展字段来满足强制要求 —— 这里显式再确认一次
        # 级别仍只来自 builder.id。
        if "x_slsaBuildLevel" in pred:
            reasons.append(
                "扩展字段 x_slsaBuildLevel=%r 被忽略：级别只能由 builder.id 决定"
                % pred["x_slsaBuildLevel"])

    return "ALLOW", reasons


# ------------------------------------------------------------------ 场景

GOOD_ARTIFACT = b"binary-v1.0.0"
GOOD_BUILDER = "https://github.com/actions/runner/github-hosted"
GOOD_EXT = {"repository": "https://github.com/octocat/hello-world",
            "ref": "refs/heads/main"}
KEY = b"demo-signing-key"


def scenarios() -> List[Tuple[str, dict, bytes, int, str]]:
    """(名称, envelope, 手上的制品, min_level, 期望)"""

    def env_of(stmt, key_id="github"):
        return sign(stmt, KEY, key_id)

    base = make_provenance(GOOD_BUILDER, "hello-world", GOOD_ARTIFACT, GOOD_EXT,
                           internal={"runnerArch": "X64"},
                           deps=[{"uri": "git+https://github.com/octocat/hello-world",
                                  "digest": {"gitCommit": "7fd1a60b01f91b314f59955a4e4d4e80d8edf11d"}}])
    out = [("S1 正常", env_of(base), GOOD_ARTIFACT, 2, "ALLOW")]

    # S2 构建后被换包
    swapped = b"binary-v1.0.0-with-backdoor"
    out.append(("S2 制品被替换", env_of(base), swapped, 2, "DENY"))

    # S3 builder 不在可接受集合里
    unknown_builder = make_provenance("https://evil.example.com/builder", "hello-world",
                                      GOOD_ARTIFACT, GOOD_EXT)
    out.append(("S3 未知 builder", env_of(unknown_builder), GOOD_ARTIFACT, 2, "DENY"))

    # S4 签名者与构建者不匹配：github 不能给 Google Cloud Build 签名
    google_builder = make_provenance("https://cloudbuild.googleapis.com/GoogleCloudBuild",
                                     "hello-world", GOOD_ARTIFACT, GOOD_EXT)
    out.append(("S4 signer-builder 不匹配", env_of(google_builder), GOOD_ARTIFACT, 2, "DENY"))

    # S5 externalParameters 被塞了意外字段（攻击者诱导的额外构建入口）
    injected = make_provenance(GOOD_BUILDER, "hello-world", GOOD_ARTIFACT,
                               dict(GOOD_EXT, entryPoint="attacker-supplied.yml"))
    out.append(("S5 externalParameters 有意外字段", env_of(injected), GOOD_ARTIFACT, 2, "DENY"))

    # S6 未知扩展字段：必须忽略，且不影响结论
    with_ext = dict(base)
    with_ext["predicate"] = dict(base["predicate"], x_customHint="ignore-me")
    out.append(("S6 未知扩展字段被忽略", env_of(with_ext), GOOD_ARTIFACT, 2, "ALLOW"))

    # S7 扩展字段自称 L3，但 builder.id 只有 L2 → 要求 L3 时必须拒绝
    lvl_ext = make_provenance(GOOD_BUILDER, "hello-world", GOOD_ARTIFACT, GOOD_EXT,
                              extra_predicate={"x_slsaBuildLevel": 3})
    out.append(("S7 扩展自称 L3 但要求 L3", env_of(lvl_ext), GOOD_ARTIFACT, 3, "DENY"))

    return out


if __name__ == "__main__":
    for name, env, art, lvl, want in scenarios():
        got, reasons = verify(env, KEY, art, lvl)
        flag = "OK " if got == want else "BAD"
        print("%s %-34s -> %-5s (want %s)" % (flag, name, got, want))
        for r in reasons:
            print("      · %s" % r)

File: 08-安全/03-应用安全与供应链/供应链完整性/provenance_verify.go
// SLSA v1.0 Provenance 的生成与验证（Go 版）。
// 重点是**验证规则**：subject 摘要绑定、signer-builder 配对、builder.id 决定级别、
// externalParameters 必须验证、internalParameters 无需验证、扩展字段必须忽略。
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"sort"
	"strings"
)

const (
	predicateType = "https://slsa.dev/provenance/v1"
	statementType = "https://in-toto.io/Statement/v1"
)

// builder.id -> 该构建平台声明的 SLSA Build level。
// 规范：builder.id "is intended to be the sole determiner of the SLSA Build level"。
var builderLevels = map[string]int{
	"github-hosted":  2,
	"cloud-build":    3,
	"evil-builder":   0,
}

// 消费者只接受特定的 signer-builder 配对。
var acceptedPairs = map[string]bool{
	"github|github-hosted": true,
	"google|cloud-build":   true,
}

// 我们对正常 externalParameters 的预期（不可信，必须下游验证）。
var expectedExternal = map[string]bool{"repository": true, "ref": true}

// Prov 是一份 provenance；未识别字段用 Extra* 显式表示，便于演示「必须忽略」。
type Prov struct {
	SubjectName    string
	SubjectDigest  string
	PredicateType  string
	BuildType      string
	External       map[string]string
	Internal       map[string]string
	Deps           []string
	BuilderID      string
	InvocationID   string
	ExtraPredicate []string // 未识别的 predicate 字段（扩展）
	ExtraTop       []string // 未识别的顶层字段
	ClaimedLevel   int      // x_slsaBuildLevel 扩展；-1 表示不存在
}

type Envelope struct {
	KeyID     string
	Signature string
	Payload   string // 规范化的 statement
}

func canonical(p Prov) string {
	var b strings.Builder
	b.WriteString(p.SubjectName + "|" + p.SubjectDigest + "|" + p.PredicateType)
	b.WriteString("|" + p.BuildType + "|" + p.BuilderID + "|" + p.InvocationID)
	keys := []string{}
	for k := range p.External {
		keys = append(keys, k)
	}
	sort.Strings(keys) // 键序无关，签名才稳定
	for _, k := range keys {
		b.WriteString("|e:" + k + "=" + p.External[k])
	}
	keys = keys[:0]
	for k := range p.Internal {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		b.WriteString("|i:" + k + "=" + p.Internal[k])
	}
	for _, d := range p.Deps {
		b.WriteString("|d:" + d)
	}
	return b.String()
}

func sign(p Prov, key []byte, keyID string) Envelope {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(canonical(p)))
	return Envelope{KeyID: keyID, Signature: hex.EncodeToString(mac.Sum(nil)),
		Payload: canonical(p)}
}

func verifySignature(e Envelope, key []byte) bool {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(e.Payload))
	return hmac.Equal(mac.Sum(nil), mustHex(e.Signature))
}

func mustHex(s string) []byte {
	b, err := hex.DecodeString(s)
	if err != nil {
		return []byte("\x00invalid")
	}
	return b
}

// verify 返回 (决定, 理由列表)
func verify(p Prov, e Envelope, key []byte, artifact []byte, minLevel int) (string, []string) {
	var reasons []string
	if !verifySignature(e, key) {
		return "DENY", []string{"签名校验失败"}
	}
	sum := sha256.Sum256(artifact)
	if p.SubjectDigest != hex.EncodeToString(sum[:]) {
		return "DENY", []string{"subject 摘要与制品不匹配（制品在构建后被替换）"}
	}
	if p.PredicateType != predicateType {
		return "DENY", []string{"predicateType 不是 SLSA provenance v1"}
	}
	// 1) signer-builder 配对
	if !acceptedPairs[e.KeyID+"|"+p.BuilderID] {
		return "DENY", []string{fmt.Sprintf(
			"不接受该 signer-builder 配对: signer=%s builder=%s", e.KeyID, p.BuilderID)}
	}
	// 2) builder.id 唯一决定级别
	level := builderLevels[p.BuilderID]
	if level < minLevel {
		msg := fmt.Sprintf("builder.id 声明的级别 %d 低于要求 %d", level, minLevel)
		if p.ClaimedLevel >= 0 {
			msg += fmt.Sprintf("；扩展字段 x_slsaBuildLevel=%d 不得用于提升级别", p.ClaimedLevel)
		}
		return "DENY", []string{msg}
	}
	// 3) externalParameters 不可信 → 拒绝未预期字段
	unexpected := []string{}
	for k := range p.External {
		if !expectedExternal[k] {
			unexpected = append(unexpected, k)
		}
	}
	if len(unexpected) > 0 {
		sort.Strings(unexpected)
		return "DENY", []string{"externalParameters 出现未预期字段: " + strings.Join(unexpected, ",")}
	}
	// 4) internalParameters 由可信平台设置 → 规范明说无需验证
	reasons = append(reasons, "internalParameters 由受信任平台设置，按规范不校验")
	// 5) 未识别字段必须忽略
	if len(p.ExtraTop) > 0 {
		reasons = append(reasons, "忽略未识别顶层字段: "+strings.Join(p.ExtraTop, ","))
	}
	if len(p.ExtraPredicate) > 0 {
		reasons = append(reasons, "忽略未识别 predicate 字段（扩展）: "+strings.Join(p.ExtraPredicate, ","))
	}
	if p.ClaimedLevel >= 0 {
		reasons = append(reasons, fmt.Sprintf(
			"扩展字段 x_slsaBuildLevel=%d 被忽略：级别只能由 builder.id 决定", p.ClaimedLevel))
	}
	return "ALLOW", reasons
}

var key = []byte("demo-signing-key")

func goodProv() Prov {
	return Prov{
		SubjectName: "hello-world",
		PredicateType: predicateType,
		BuildType:    "https://slsa-framework.github.io/github-actions-buildtypes/workflow/v1",
		External: map[string]string{
			"repository": "https://github.com/octocat/hello-world",
			"ref":        "refs/heads/main"},
		Internal:     map[string]string{"runnerArch": "X64"},
		Deps:         []string{"gitCommit:7fd1a60b01f91b314f59955a4e4d4e80d8edf11d"},
		BuilderID:    "github-hosted",
		InvocationID: "inv-0001",
		ClaimedLevel: -1,
	}
}

type scenario struct {
	name     string
	prov     Prov
	sigKeyID string
	artifact []byte
	minLevel int
	want     string
}

func scenarios() []scenario {
	artifact := []byte("binary-v1.0.0")
	base := goodProv()
	base.SubjectDigest = digestOf(artifact)
	out := []scenario{
		{"S1 正常", base, "github", artifact, 2, "ALLOW"},
		{"S2 制品被替换", base, "github", []byte("binary-v1.0.0-with-backdoor"), 2, "DENY"},
	}
	evil := goodProv()
	evil.BuilderID = "evil-builder"
	evil.SubjectDigest = digestOf(artifact)
	out = append(out, scenario{"S3 未知 builder", evil, "github", artifact, 2, "DENY"})

	gcb := goodProv()
	gcb.BuilderID = "cloud-build"
	gcb.SubjectDigest = digestOf(artifact)
	out = append(out, scenario{"S4 signer-builder 不匹配", gcb, "github", artifact, 2, "DENY"})

	inj := goodProv()
	inj.SubjectDigest = digestOf(artifact)
	inj.External = map[string]string{
		"repository": base.External["repository"], "ref": base.External["ref"],
		"entryPoint": "attacker-supplied.yml"}
	out = append(out, scenario{"S5 externalParameters 有意外字段", inj, "github", artifact, 2, "DENY"})

	withExt := goodProv()
	withExt.SubjectDigest = digestOf(artifact)
	withExt.ExtraPredicate = []string{"x_customHint"}
	out = append(out, scenario{"S6 未知扩展字段被忽略", withExt, "github", artifact, 2, "ALLOW"})

	lvl := goodProv()
	lvl.SubjectDigest = digestOf(artifact)
	lvl.ExtraPredicate = []string{"x_slsaBuildLevel"}
	lvl.ClaimedLevel = 3
	out = append(out, scenario{"S7 扩展自称 L3 但要求 L3", lvl, "github", artifact, 3, "DENY"})
	return out
}

func digestOf(b []byte) string {
	s := sha256.Sum256(b)
	return hex.EncodeToString(s[:])
}

func main() {
	fail := 0
	for _, sc := range scenarios() {
		e := sign(sc.prov, key, sc.sigKeyID)
		got, reasons := verify(sc.prov, e, key, sc.artifact, sc.minLevel)
		flag := "OK "
		if got != sc.want {
			flag = "BAD"
			fail++
		}
		fmt.Printf("%s %-34s -> %-5s (want %s)\n", flag, sc.name, got, sc.want)
		for _, r := range reasons {
			fmt.Printf("      · %s\n", r)
		}
	}
	// 同一份证明，把要求级别从 2 提到 3 → 结论反转
	artifact := []byte("binary-v1.0.0")
	g := goodProv()
	g.SubjectDigest = digestOf(artifact)
	e := sign(g, key, "github")
	if got, _ := verify(g, e, key, artifact, 2); got != "ALLOW" {
		fmt.Println("FAIL 要求 L2 应通过")
		fail++
	}
	if got, _ := verify(g, e, key, artifact, 3); got != "DENY" {
		fmt.Println("FAIL 要求 L3 应拒绝")
		fail++
	}
	if verifySignature(Envelope{KeyID: "github", Signature: "deadbeef",
		Payload: canonical(g)}, key) {
		fmt.Println("FAIL 伪造签名不应通过")
		fail++
	}
	if fail > 0 {
		fmt.Printf("FAILED %d\n", fail)
		os.Exit(1)
	}
	fmt.Println("all checks passed")
}
